sdk/helper/keysutil/policy.go

```diff
@@ -244,6 +244,10 @@ type KeyEntry struct {
 	DeprecatedCreationTime int64 `json:"creation_time"`
 
 	ManagedKeyUUID string `json:"managed_key_id,omitempty"`
+
+	// Key entry certificate chain. If set, leaf certificate key matches the
+	// KeyEntry key
+	CertificateChain []*x509.Certificate `json:"certificate_chain"`
```

Comment:
Well, I have made this change but I'm not sure if I understood what you meant with the "how will marshal and unmarshal" work. would you mind expanding?

Reply:
Ah sorry, forgot to expand on this before merging. I had a great demo of this last week, wish I had remember you asked this question and I would've put it here :laughing: 

Let's take two certificates: root and leaf. If we take the root certificate and round-trip it through JSON, you'll see that the verification of the leaf fails, even though it verifies fine without the JSON.

<details>

<summary> Go code and execution </summary>

Go code example: 

```go
// You can edit this code!
// Click here and start typing.
package main

import (
	"crypto/x509"
	"encoding/json"
	"encoding/pem"
	"fmt"
)

const rootPEM = `-----BEGIN CERTIFICATE-----
MIIBnzCCAUWgAwIBAgIUYBcTTOiYMhGTyfu8NRbSzUZHmdswCgYIKoZIzj0EAwIw
HTEbMBkGA1UEAxMSTG9uZy1MaXZlZCBSb290IFgxMB4XDTIzMDgyMjEyMjYyMVoX
DTIzMDkyMzEyMjY1MVowHTEbMBkGA1UEAxMSTG9uZy1MaXZlZCBSb290IFgxMFkw
EwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE3abeYSwv7MvVf7zpD2QJ//9iuaf5bn0U
E4ZdBsp94OcFo8CwflTR0CYQywJ68wzcClvAYg4ZkYbTT+Z3xVDRS6NjMGEwDgYD
VR0PAQH/BAQDAgEGMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFGpFtIseCP0r
RORRpB6rz08D+/ERMB8GA1UdIwQYMBaAFGpFtIseCP0rRORRpB6rz08D+/ERMAoG
CCqGSM49BAMCA0gAMEUCIQCHiXQp7fzkjZ+Ls4Ez3m8Z/zowvaKPFX03JosNxkCt
CQIgKBbbuugqm77apjHteC2tm9zHvm1BseU5BJMqOCEpv8M=
-----END CERTIFICATE-----
`
const leafPEM = `-----BEGIN CERTIFICATE-----
MIIBtzCCAV6gAwIBAgIUFfjLDOAHeo2Tns1987cYgULUysswCgYIKoZIzj0EAwIw
HTEbMBkGA1UEAxMSTG9uZy1MaXZlZCBSb290IFgxMB4XDTIzMDgyMjEyMjY1M1oX
DTIzMDgyMjEyMjcyOFowEjEQMA4GA1UEAxMHdGVzdGluZzBZMBMGByqGSM49AgEG
CCqGSM49AwEHA0IABMTjizFLVEVsCNhJbK/m565uHqsgh5wFtRZ8mJNN2L2Yg0XG
PEN/D5YiLDcLE7WTAAm2mrt1TrqdOEGOripdU+CjgYYwgYMwDgYDVR0PAQH/BAQD
AgOoMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAdBgNVHQ4EFgQUPmoJ
10pHYZMsuhRMl5Q34327d84wHwYDVR0jBBgwFoAUakW0ix4I/StE5FGkHqvPTwP7
8REwEgYDVR0RBAswCYIHdGVzdGluZzAKBggqhkjOPQQDAgNHADBEAiBrqAj9ocjR
XmzpvWmGQBxa25fNTEMWYTaCEAy8VkZRYgIgT3dyVXQGjffQyWF4o0KLNOAUH3Om
GA30tvXYKZ/8ksU=
-----END CERTIFICATE-----
`

func main() {
	rootBlock, _ := pem.Decode([]byte(rootPEM))
	rootOrig, _ := x509.ParseCertificate(rootBlock.Bytes)

	leafBlock, _ := pem.Decode([]byte(leafPEM))
	leafOrig, _ := x509.ParseCertificate(leafBlock.Bytes)

	if err := leafOrig.CheckSignatureFrom(rootOrig); err != nil {
		fmt.Printf("failed to validate original signature: %v\n", err)
	} else {
		fmt.Printf("checked original signatures: OK\n")
	}

	// XXX: this is what we were doing prior to your
	// change to [][]byte: we were storing it as a JSON-encoded
	// struct field of type *x509.Certificate -- that when stored
	// (marshalled) and read again later (unmarshalled), would
	// not be a "functioning" certificate.
	rootCertStored, _ := json.Marshal(rootOrig)
	var rootUnstored *x509.Certificate
	json.Unmarshal(rootCertStored, &rootUnstored)

	if err := leafOrig.CheckSignatureFrom(rootUnstored); err != nil {
		fmt.Printf("failed to validate round-tripped signature: %v\n", err)
		fmt.Printf("unstored root key: (type: %T) %v\n", rootUnstored.PublicKey, rootUnstored.PublicKey)
		fmt.Printf("orig root key: (type: %T) %v\n", rootOrig.PublicKey, rootOrig.PublicKey)
	} else {
		fmt.Printf("checked round-tripped signatures: OK\n")
	}
}
```

Execution results:

```
checked original signatures: OK
failed to validate round-tripped signature: x509: cannot verify signature: algorithm unimplemented
unstored root key: (type: map[string]interface {}) map[Curve:map[] X:1.0025597095711236e+77 Y:2.550890208028647e+75]
orig root key: (type: *ecdsa.PublicKey) &{0x685500 100255970957112353876818454842956629422432050733395608904658668741126008987879 2550890208028647296912217307641692388690570932323450769480190751495358042443}

```

</details>

https://go.dev/play/p/zX9sS8M8vk1

IOW, because (after unmarshaling the certificate) you get different types for the public key (it is an untyped `interface{}` after all), you can't round-trip Go's x509 certificate structure via JSON, like we do for storage here. So you're best off using a round-tripping type, so you don't have to deal with how do you "fix" the unmarshal if you want to do any more advanced operations on it in the future (e.g., verify signatures using the certificate in Vault -- for chain building/verification for instance).